workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/SpatialOSBehaviourManager.cs

```diff
@@ -1,60 +1,202 @@
+            foreach (var behaviour in spatialOSBehavioursToDisable)
+            {
+                behaviour.enabled = false;
+                behaviourLibrary.DeInjectAllReadersWriters(behaviour);
+                foreach (var idToReaderWriter in behaviourToCreatedReadersWriters[behaviour])
+                {
+                    var id = idToReaderWriter.Key;
+                    var reader = idToReaderWriter.Value;
+                    componentIdToCreatedReadersWriters[id].Remove(reader);
+                }
+
+                behaviourToCreatedReadersWriters.Remove(behaviour);
+            }
+
+            spatialOSBehavioursToDisable.Clear();
         }
 
         public void AddComponent(uint componentId)
         {
-            // Mark readers as to be enabled
+            // Mark reader components ready in relevant SpatialOSBehaviours
+            var relevantReaderSpatialOSBehaviours = componentIdToReaderSpatialOSBehaviours[componentId];
+            foreach (var spatialOSBehaviour in relevantReaderSpatialOSBehaviours)
+            {
+                MarkComponentReady(spatialOSBehaviour);
+            }
         }
 
         public void RemoveComponent(uint componentId)
         {
-            // Mark readers as to be disabled
+            // Mark reader components not ready in relevant SpatialOSBehaviours
+            var relevantReaderSpatialOSBehaviours = componentIdToReaderSpatialOSBehaviours[componentId];
+            foreach (var spatialOSBehaviour in relevantReaderSpatialOSBehaviours)
+            {
+                MarkComponentNotReady(spatialOSBehaviour);
+            }
         }
 
         public void ChangeAuthority(uint componentId, Authority authority)
         {
-            // Mark writers as to be enabled/disabled
+            if (authority == Authority.Authoritative)
+            {
+                // Mark writer components ready in relevant SpatialOSBehaviours
+                var relevantWriterSpatialOSBehaviours = componentIdToWriterSpatialOSBehaviours[componentId];
+                foreach (var spatialOSBehaviour in relevantWriterSpatialOSBehaviours)
+                {
+                    MarkComponentReady(spatialOSBehaviour);
+                }
+            }
+            else if (authority == Authority.NotAuthoritative)
+            {
+                // Mark writer components not ready in relevant SpatialOSBehaviours
+                var relevantWriterSpatialOSBehaviours = componentIdToWriterSpatialOSBehaviours[componentId];
+                foreach (var spatialOSBehaviour in relevantWriterSpatialOSBehaviours)
+                {
+                    MarkComponentNotReady(spatialOSBehaviour);
+                }
+            }
+        }
+
+        private void MarkComponentReady(MonoBehaviour spatialOSBehaviour)
+        {
+            // Inject all Readers/Writers at once when all requirements are met
+            numUnsatisfiedReadersOrWriters[spatialOSBehaviour]--;
```

Comment:
Does this actually decrease the value stored?

Reply:
Yes (see link on Jamie's comment above)